compiler/rustc_hir_typeck/src/pat.rs

```diff
@@ -2039,9 +2135,19 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         slice: Option<&'tcx Pat<'tcx>>,
         after: &'tcx [Pat<'tcx>],
         expected: Ty<'tcx>,
-        def_bm: BindingMode,
-        ti: TopInfo<'tcx>,
+        pat_info: PatInfo<'tcx, '_>,
     ) -> Ty<'tcx> {
+        // If the pattern is irrefutable and `expected` is an infer ty, we try to equate it
+        // to an array if the given pattern allows it. See issue #76342
+        if self.pat_is_irrefutable(pat_info.decl_origin) && expected.is_ty_var() {
```

Comment:
please move this `if` until after the `structurally_resolve_type` as it can otherwise miss info about the type of expected.

Reply:
The new tests previously failed because `structurally_resolve_type` returned an inference error.

Reply:
I'm not convinced this is necessary. We only get into the `demand_eqtype` call when the pattern is irrefutable, how exactly would trying to structurally resolve the type help here? Afaict either there's a type annotation and we ignore all the logic introduced in this PR or there is none, in which case `expected` is just an unresolvable type variable afaict. Since we have to resolve the type after the `demand_eqtype` call anyway (since we still have to resolve the inner type after hinting that we're looking for `[_; n]`) I don't see why we want `try_structurally_resolve_type` first. Can you explain why that would be helpful/necessary?

Reply:
we get into the `demand_eqtype` call when the pattern is irrefutable **and** the expected type is an inference variable. Whether the expected type is an inference variable can depend on structrally resolving it.

I can't imagine how you'd get to a point where it matters, but generally looking at the `kind()` of a type and then using `structurally_resolve` or `shallow_resolve` or whatever  is always sus. Either the `resolve` is a noop, in which case it should be removed, or it can change the `kind()` of the type, in which case looking at the type before the resovle call is buggy.

Ah, nearly forgot: can you add tests for destructing assignments, i.e. what happens with the following
```rust
fn main() {
    let a: i32;
    let b;
    [a, b] = Default::default();
}
```

we should treat these as also being irrefutable and this example should compile i think